src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java

```diff
@@ -161,130 +237,766 @@ public ResultSet getTableTypes() throws SQLException {
+                new BsonElement(IS_AUTOINCREMENT, new BsonString("NO")),
+                new BsonElement(IS_GENERATEDCOLUMN, new BsonString("")));
+    }
+
+    // Helper for creating BSON documents for the getColumnPrivileges methods. Intended
+    // for use with the getColumnsFromDB helper method which is shared between getColumns
+    // and getColumnPrivileges.
+    private BsonDocument toGetColumnPrivilegesDoc(GetColumnsDocInfo i) {
+        return createSortableBottomBson(
+                // Per JDBC spec, sort by  COLUMN_NAME and PRIVILEGE. Since all PRIVILEGEs are the same,
+                // we just sort by COLUMN_NAME here.
+                GET_COLUMN_PRIVILEGES_SORT_SPECS,
+                new BsonElement(TABLE_CAT, new BsonString(i.dbName)),
+                new BsonElement(TABLE_SCHEM, new BsonString("")),
+                new BsonElement(TABLE_NAME, new BsonString(i.tableName)),
+                new BsonElement(COLUMN_NAME, new BsonString(i.columnName)),
+                new BsonElement(GRANTOR, BsonNull.VALUE),
+                new BsonElement(GRANTEE, new BsonString("")),
+                new BsonElement(PRIVILEGE, new BsonString("SELECT")),
+                new BsonElement(IS_GRANTABLE, BsonNull.VALUE));
+    }
+
+    // Helper for ensuring a sqlGetSchema result is a valid collection schema. As in,
+    // it has ok: 1, has a jsonSchema, and the jsonSchema is an object schema.
+    private boolean isValidSchema(MongoJsonSchemaResult res) {
+        return res.ok == 1 && res.schema.jsonSchema != null && res.schema.jsonSchema.isObject();
+    }
+
+    // Helper for getting column data for all columns from all tables from a specific
+    // database. Used by getColumns and getColumnPrivileges. The caller specifies how
+    // to serialize the column info into BSON documents for the result set.
+    private Stream<BsonDocument> getColumnsFromDB(
+            String dbName,
+            Pattern tableNamePatternRE,
+            Pattern columnNamePatternRE,
+            Function<GetColumnsDocInfo, BsonDocument> bsonSerializer) {
+        MongoDatabase db = this.conn.getDatabase(dbName).withCodecRegistry(MongoDriver.registry);
+
+        return db.listCollectionNames()
+                .into(new ArrayList<>())
+                .stream()
+
+                // filter only for collections matching the pattern
+                .filter(tableName -> tableNamePatternRE.matcher(tableName).matches())
+
+                // map the collection names into triples of (dbName, tableName, tableSchema)
+                .map(
+                        tableName ->
+                                new Pair<>(
+                                        new Pair<>(dbName, tableName),
+                                        db.runCommand(
+                                                new BsonDocument(
+                                                        "sqlGetSchema", new BsonString(tableName)),
```

Comment:
Is this command correct: `sqlGetSchema`?  In MongoSQLStatement the command used is `sqlGetResultSchema`.

Reply:
Yes, `sqlGetSchema` ([docs](https://docs.mongodb.com/datalake/reference/cli/sql/sqlgetschema/)) is the command used to get the schema for a _collection_. The command you used in the other PR is `sqlGetResultSchema` which is used to get the result set schema for a _query_. Here, we need the collection schema so we could get type information for the fields.